### ISSUE TITLE
Add a `ready: boolean` to `Router` returned by `useRouter`

### ISSUE THREAD

**Issue by jtomaszewski** [vercel/next.js#8259]
Because of the way Next.js does initial render of the pages, and that the initial render on client-side and server-side has to be the same (see https://reactjs.org/docs/react-dom.html#hydrate ), sometimes when a page is rendered, its' `query` (returned i.e. by `useRouter()` hook) will be empty, even though it actually has query params. After the initial render, soon after there will be another render in which the `query` will be then populated with proper query params. ( Example described also in i.e. https://github.com/zeit/next.js/issues/7880 ) 

My code relies on those query params very heavily and we need to know "are those the real current query params or not yet?". (i.e. because if query params are missing or incorrect, we'd like to redirect the user to another page)

## Describe the solution you'd like

I think that either:

- `query` should be always populated with the current query params (even on initial client-side hydration that happens after a server-side render)
- or, (IMO better), add a `ready: boolean` flag to `Router`, so the component can listen to this using i.e. `useEffect` and ignore the query params when they aren't parsed yet.

## Describe alternatives you've considered

I think we'll have to go with some `setTimeout` or sth for now until a better solution is found.

**Comment by lfades** [comment]
`query` should be available in `useEffect` because it's only used in the browser, your page is most likely being prerendered and therefore the query params are empty until the browser does the render.

You can either keep all the changes that depend in the query params inside `useState` or add `getInitialProps` if you need to have the query params always defined, in exchange of not having a statically exported page

**Comment by jtomaszewski** [comment]
> query should be available in useEffect because it's only used in the browser, your page is most likely being prerendered and therefore the query params are empty until the browser does the render.

I wish that was true... but it isn't. Besides `useEffect` there also needs to be a `setTimeout`. Only after that the route params are finally there.

For now we have a workaround like the following, but ideally, I'd prefer `next/router` to provide that out-of-the-box:

```ts
import NextRouter, {
  useRouter as useNextRouter,
  NextRouter,
} from 'next/router';

const RouterContext = React.createContext<NextRouter & { ready: boolean }>(
  null as any,
);
export function RouterContextProvider({
  children,
}: {
  children?: React.ReactNode;
}) {
  const router = useNextRouter();

  // In order to know if the params that are passed by router
  // are actually correct
  // (because sometimes on initial render they might be equal to server-side params,
  //  not the ones that are currently visible in the browser),
  // we'll be relying on this `ready: boolean` property.
  // ( Workaround until next.js give us a proper solution for it
  //  - see https://github.com/zeit/next.js/issues/8259 )
  const [ready, setReady] = useState(false);
  useEffect(() => {
    const timeout = setTimeout(() => {
      setReady(true);
    });
    return () => clearTimeout(timeout);
  }, []);

  const wrappedRouter = useMemo(
    () => ({
      ...router,
      ready,
    }),
    [router, ready],
  );

  return (
    <RouterContext.Provider value={wrappedRouter}>
      {children}
    </RouterContext.Provider>
  );
}

export function useRouter() {
  return useContext(RouterContext);
}
```

> You can either keep all the changes that depend in the query params inside useState or add getInitialProps if you need to have the query params always defined, in exchange of not having a statically exported page

We can't use `getInitialProps` - those query params aren't known at the time of server-side render. Also, we want that page to be exported (we actually explicitly call `next export`, because that's the way our app has to be built like).

**Comment by Timer** [comment]
Sorry, but you cannot branch rendering logic based on booleans -- this causes hydration mismatches and typically indicates a code pattern that needs fixed.

> My code relies on those query params very heavily and we need to know "are those the real current query params or not yet?"

You can rely directly on `location.search` within `useEffect` if you cannot tolerate Next.js informing you of new query parameters in the way it does.

Also, this `setTimeout` trick provides no guarantees the updated query object is going to be to you.

**Comment by jtomaszewski** [comment]
> Sorry, but you cannot branch rendering logic based on booleans -- this causes hydration mismatches and typically indicates a code pattern that needs fixed.

Yeah, but the RouterContext is updated once the query params is updated, right? So together with it, it update the `ready` property on that context, after the client updates the params.

I want to avoid branching rendering logic based on "am I on ssr?" booleans to make hydration work, thus raised this issue.

> You can rely directly on location.search within useEffect if you cannot tolerate Next.js informing you of new query parameters in the way it does.

We have routes with dynamic parameters. Parsing `location.search` on our own would be troublesome and would be a duplication of what Next router does.

**Comment by jtomaszewski** [comment]
Our example is like this:

- We have a route like `/rooms`.
- Our app is served using CDN, through `next export`
- That `/rooms` accepts multiple query params like `?region=us&property=hotel`. At least one of them needs to be present. If none of them are present or if they are invalid, we want to show the user an error (or redirect him to a 404 page).

So basically, we need to be able get the initial route params on/after the page loads, and then act on it (i.e. show an error to the user if the params are incorrect).

However, currently, without the above workaround, if we navigated to the `/rooms?region=us`:

1. On page load, the route params are empty.
2. After a certain moment (let's say the `useEffect` + `setTimeout`), the route params get populated with actual params that are visible at the client-side.

If we won't rely on sth like `router.ready`, then if we show error at the 1st moment, then it will be incorrect UI behaviour: first there would be an error visible ("your URL is wrong!"), and a subsequent render after that, the error would be hidden, because the component would then assume that URL is actually correct.

We'd like to avoid having weird workarounds like surrounding the "show error code" with `setTimeout(..., 500)`. Instead, it would be perfect to know if the route params are actually the params that are visible on client-side, or if they are just the old version from server-side (because the client-side query params aren't known yet).

**Comment by iwarner** [comment]
Hiya I ended up putting the query into state:

~~~
  const [bankId, setBankId] = useState(null)
  const { query } = useRouter()

  useEffect(() => {
    setBankId(query.bankId)

    if (bankId) {
      const filesFiltered = filter(files, (o) => { return o.dir === bankId })
      setFiles(filesFiltered)
      setIsLoading(false)
    }
  }, [query, bankId])
~~~

This cause multiples render cycles, but I do not see any other way of getting the router params.

I am using a Static build, so do not want to use getInitial... be great if examples could showcase static methods also to state best practices.

**Comment by defx** [comment]
+1 for better documentation around static patterns.

**Comment by bmvantunes** [comment]
I was having exactly the same problem - I solved it the following way:

@jtomaszewski you can have something like this:
```typescript
import { NextRouter } from 'next/router';

export function isRouterReady(router: NextRouter) {
  return router.asPath !== router.route;
}
```

If you are using pre-rendered pages, immediately after hydration the router.route property and router.asPath property will have exactly the same value.

In your case, in the first render you'll have:
```
{
  asPath: "/rooms"
  router: "/rooms"
}
```

In the second render you'll have something like:
```
{
  asPath: "/rooms?myProperty=something"
  router: "/rooms"
}
```

**Comment by Timer** [comment]
Upon further thought, I don't see the harm in supporting `Router.ready`. Feel free to send a PR!

**Comment by Soundvessel** [comment]
I ran into a similar issue where I am using the `useRouter` hook inside a custom hook used for several components to update query params (table column sorting, filter UI, all param based). The first component of four would always have a stale query value reported at the `console.log`. This occurred even after several interactions with the other components.

```
import { useCallback } from 'react'
import { useRouter } from 'next/router'
import { Router } from '../routes'
import { getRouteNameFromAsPath } from '../lib/utils'
export default function usePushQueryParams() {
  const { asPath, query } = useRouter()
  const pushQueryParams = useCallback(
    function pushQueryParamsUpdate(newQueryParams) {
      console.log('useRouter Query', query)
      const routeName = getRouteNameFromAsPath(asPath)
      const newQuery = {
        ...query,
        ...newQueryParams,
      }
      Router.pushRoute(routeName, newQuery)
    },
    [asPath, query]
  )
  return pushQueryParams
}
```

It is rather confusing in these instances, where hooks should be compostable in custom hooks, that `useRouter` doesn't reliably return the current query. Perhaps an update to the documentation can at least provide others a sanity check?

**Comment by lfades** [comment]
@Soundvessel In the case of prerendered pages (pages without `getInitialProps`), the `query` object will be empty in the first render, that will get populated after client-side hydration. A place that always has access to a populated query object is `useEffect`, so try accessing to it from there.

I agree that docs don't currently do a good job explaining it, but it will get fixed soon 👍

**Comment by Soundvessel** [comment]
Ah and I think a ref is a good way to set something in that `useEffect`?

Here is my now working updated hook.

```
import { useCallback, useEffect, useRef } from 'react'
import { useRouter } from 'next/router'
import { Router } from '../routes'
import { getRouteNameFromAsPath } from '../lib/utils'

export default function usePushQueryParams() {
  const { asPath, query } = useRouter()

  const queryRef = useRef(null)

  useEffect(() => {
    queryRef.current = query
  }, [query])

  const pushQueryParams = useCallback(
    function pushQueryParamsUpdate(newQueryParams) {
      const routeName = getRouteNameFromAsPath(asPath)
      const newQuery = {
        ...queryRef.current,
        ...newQueryParams,
      }
      // eslint-disable-next-line no-console
      Router.pushRoute(routeName, newQuery)
    },
    [asPath]
  )

  return pushQueryParams
}
```

**Comment by Soundvessel** [comment]
What is rather weird is that I made a custom hook since I needed to use the query in several other components. For some reason the above works but when I switch to it using this hook which uses the same code I get the stale query again. Looks like the useEffect needs to exist where you are using the query and can't be from a custom hook?

```
export default function useUrlQuery() {
  const { query } = useRouter()

  const queryRef = useRef(query)

  useEffect(
    function setHydratedQuery() {
      queryRef.current = query
    },
    [query]
  )

  return queryRef.current
```

**Comment by julian-sf** [comment]
> I was having exactly the same problem - I solved it the following way:
> 
> @jtomaszewski you can have something like this:
> 
> ```ts
> import { NextRouter } from 'next/router';
> 
> export function isRouterReady(router: NextRouter) {
>   return router.asPath !== router.route;
> }
> ```
> 
> If you are using pre-rendered pages, immediately after hydration the router.route property and router.asPath property will have exactly the same value.
> 
> In your case, in the first render you'll have:
> 
> ```
> {
>   asPath: "/rooms"
>   router: "/rooms"
> }
> ```
> 
> In the second render you'll have something like:
> 
> ```
> {
>   asPath: "/rooms?myProperty=something"
>   router: "/rooms"
> }
> ```
> 
> @Timer can you confirm if this is a valid approach?

This only works if you expect query params to be populated.  IE, if there were NO query params to be added, then this would never return true, which is a bug.

I think the only clean way to "know" at least from client code, is to have the query params in the `useEffect` hook.  I'm going to look through the next code and see if there is a clean way to decipher this "ready" state.  It is essentially tied to whether or not you are "post" hydration, aka `ready === haveIHydratedYet`.

**Comment by julian-sf** [comment]
Made a PR that adds a new parameter called `queryReady` here: #9370 

I would love some feedback on the name as it would look like this in practice:

```typescript
const { query, queryReady } = useRouter();
```

where `queryReady` is false until `query` has been populated.

I HAD to use something that wasn't `ready` as `ready` was already defined on the router singleton here:

```typescript
type SingletonRouterBase = {
  router: Router | null
  readyCallbacks: Array<() => any>
  ready(cb: () => any): void
}
```

Please see the PR for more detailed reasoning...

**Comment by sidujjain** [comment]
> > I was having exactly the same problem - I solved it the following way:
> > @jtomaszewski you can have something like this:
> > ```ts
> > import { NextRouter } from 'next/router';
> > 
> > export function isRouterReady(router: NextRouter) {
> >   return router.asPath !== router.route;
> > }
> > ```
> > 
> > 
> > If you are using pre-rendered pages, immediately after hydration the router.route property and router.asPath property will have exactly the same value.
> > In your case, in the first render you'll have:
> > ```
> > {
> >   asPath: "/rooms"
> >   router: "/rooms"
> > }
> > ```
> > 
> > 
> > In the second render you'll have something like:
> > ```
> > {
> >   asPath: "/rooms?myProperty=something"
> >   router: "/rooms"
> > }
> > ```
> > 
> > 
> > @Timer can you confirm if this is a valid approach?
> 
> This only works if you expect query params to be populated. IE, if there were NO query params to be added, then this would never return true, which is a bug.
> 
> I think the only clean way to "know" at least from client code, is to have the query params in the `useEffect` hook. I'm going to look through the next code and see if there is a clean way to decipher this "ready" state. It is essentially tied to whether or not you are "post" hydration, aka `ready === haveIHydratedYet`.

I'm seeing multiple rendering issue only with dynamic routes. This seems to work:
```
import { isDynamicRoute } from "next/dist/next-server/lib/router/utils/is-dynamic";

export function isRouterReady(router) {
  return !isDynamicRoute(router.pathname) || router.asPath !== router.route;
}
```

**Comment by wilsonpage** [comment]
Running into this issue also. From my ignorant next internals perspective it seems strange that `router.asPath` can't mirror `window.location` on first render for static pages. `router.ready` would be an acceptable workaround, but it's still a gotcha that nextjs user will need to be aware of.

I'm using @bmvantunes's approach for now:

```js
const testRouterReady = (router) => router.asPath !== router.route;
```

**Comment by urlund** [comment]
+1

**Comment by ro0sterjam** [comment]
Thanks for the detailed write up! Recently ran into this issue when I noticed my component was making two api calls, one with the query param populated and one without. I'm surprised this issue is not more common.

Was just wondering if there was any updates to this pr? Is there currently a generally accepted work around in the meantime?

**Comment by rafinskipg** [comment]
+1

**Comment by rafinskipg** [comment]
This issue is also happening with dynamic paths. Having `/[slug]` on the path instead of the real route

**Comment by dmytro-lymarenko** [comment]
> Running into this issue also. From my ignorant next internals perspective it seems strange that `router.asPath` can't mirror `window.location` on first render for static pages. `router.ready` would be an acceptable workaround, but it's still a gotcha that nextjs user's will need to be aware of.
> 
> I'm using similar to @bmvantunes's approach for now:
> 
> ```js
> const isDynamicPage = (router) => /\[.+\]/.test(router.route);
> const testRouterReady = (router) => !isDynamicPage(router) || router.asPath !== router.route;
> ```

For me it doesn't work in all cases. So, I use something like this:
```ts
const isDynamicPage = (router: NextRouter) => /\[.+\]/.test(router.route);
const testRouterReady = (router: NextRouter) =>
	isDynamicPage(router) || router.asPath.includes('?') ? Object.keys(router.query).length > 0 : true;
```
which means if route is dynamic or asPath contains some query then `query` object should have some values

**Comment by lifeiscontent** [comment]
@jtomaszewski would you ever want to use the router when it is not "ready"? I feel like next.js should only render a page once the router is ready, that way I don't have to do a bunch of conditional logic around if query params are loaded or not.

like in this example:

https://github.com/lifeiscontent/realworld/blob/2feda02ec18d617b4806dad071d8a97908f52e0d/web/src/containers/index-page/index.js#L14-L25

to me, it'd be ideal if next would just wait for the params to be ready before rendering, that way I don't need to conditionally check if they're "ready" which seems like an implementation detail to me that shouldn't be exposed to us as the users of next.js

What are your thoughts?

**Comment by lifeiscontent** [comment]
CC: @julian-sf ^

**Comment by timneutkens** [comment]
> would you ever want to use the router when it is not "ready"? I feel like next.js should only render a page once the router is ready, that way I don't have to do a bunch of conditional logic around if query params are loaded or not.

> to me, it'd be ideal if next would just wait for the params to be ready before rendering, that way I don't need to conditionally check if they're "ready" which seems like an implementation detail to me that shouldn't be exposed to us as the users of next.js

That would break hydration and cause the whole DOM tree to be thrown away.

**Comment by jtomaszewski** [comment]
> @jtomaszewski would you ever want to use the router when it is not "ready"?

Let me reframe your question: Would you want your user to be able to see the page even when the client script isn't initialized yet (and client route params aren't known yet)? 

The answer is of course yes. You need to do that conditional logic because there needs to be a server-side response and a separate client-side one once the route params are known.

I agree that the logic `params = router.ready ? router.params : undefined` isn't the nicest one. Another way would be to have for example `router.serverParams: Params` or `router.initialParams: Params` for the initial params provided by the server, and `router.params: Params | undefined` for the other cases.

**Comment by lifeiscontent** [comment]
@timneutkens sorry for my naivety, can you elaborate on that a bit more just so I can have a better understanding of how it works?

@jtomaszewski I guess I don't understand what you're getting at, the params come from the URL which is available on both the server and the client. Or are you specifically talking about params that might be from dynamic route segments?

**Comment by timneutkens** [comment]
@lifeiscontent https://reactjs.org/docs/react-dom.html#hydrate

When the React tree does not match 1-1 it throws a hydration warning and discards all pre-rendered DOM of that tree causing a full re-render of the DOM that is already rendered.

When pre-rendering to static html (eg through automatic static optimization) the dynamic route segments / querystring values are not available as the served response is static. This means that if the client-side JS boots up it has to mirror the tree that was there when pre-rendering, without the querystring values, otherwise React throws away the DOM tree. While running hydrate `useEffect` is called (it only runs in the browser) and that's when we do the querystring / dynamic route parsing.

Dynamic route parsing only happens client-side when it's automatic static optimization btw, as with getStaticProps you provide getStaticPaths also

**Comment by lifeiscontent** [comment]
@timneutkens thanks for the detailed explanation. That helped paint a much clearer picture for me.

**Comment by daveteu** [comment]
> > I was having exactly the same problem - I solved it the following way:
> > @jtomaszewski you can have something like this:
> > ```ts
> > import { NextRouter } from 'next/router';
> > 
> > export function isRouterReady(router: NextRouter) {
> >   return router.asPath !== router.route;
> > }
> > ```
> > 
> > 
> > If you are using pre-rendered pages, immediately after hydration the router.route property and router.asPath property will have exactly the same value.
> > In your case, in the first render you'll have:
> > ```
> > {
> >   asPath: "/rooms"
> >   router: "/rooms"
> > }
> > ```
> > 
> > 
> > In the second render you'll have something like:
> > ```
> > {
> >   asPath: "/rooms?myProperty=something"
> >   router: "/rooms"
> > }
> > ```
> > 
> > 
> > @Timer can you confirm if this is a valid approach?
> 
> This only works if you expect query params to be populated. IE, if there were NO query params to be added, then this would never return true, which is a bug.
> 
> I think the only clean way to "know" at least from client code, is to have the query params in the `useEffect` hook. I'm going to look through the next code and see if there is a clean way to decipher this "ready" state. It is essentially tied to whether or not you are "post" hydration, aka `ready === haveIHydratedYet`.

**I've tried the above mentioned method and tried comparing the asPath and my results seems different from what was mentioned. My asPath includes the parameter.**

```
Initial Load
router.asPath – "/test?this=is&a=parameter"
router.path + query – "/test?"
router.query – {} 

subsequent render
router.asPath – "/test?this=is&a=parameter"
router.path + query – "/test?this=is&a=parameter"
router.query – {this: "is", a: "parameter"}
```

**Simple code I used to test.**
```
import {useRouter} from 'next/router'
import { useEffect } from 'react';
var qs = require('querystring')
const Test = () => {
  const router = useRouter();
  useEffect(() => {  
    console.log('router.asPath', router.asPath);
    console.log('router.path + query', router.route+'?'+qs.encode(router.query))
    console.log('router.query', router.query);
  })  
  return ( <h1>test</h1>)
}
export default Test;
```
**The results as follows** 
![image](https://user-images.githubusercontent.com/18649392/81771516-cc243c00-9515-11ea-9340-b54b285b9638.png)

Hence I believe comparing asPath and path+query should be sufficient to check if router is ready for the use case. If the 2 strings is the same, we can go ahead with our first fetch from the api server.

Let me know your thoughts.

**Comment by lifeiscontent** [comment]
@timneutkens any idea when this might be merged?

**Comment by dmytro-lymarenko** [comment]
These two strings `/test?this=is&a=parameter` and `/test?a=parameter&this=is` are not equal but they contain the same information

**Comment by bduffany** [comment]
Here's a workaround which combines some of the ideas in this thread, fixing some of the problems and edge cases while also providing a clean API:

```typescript
// useQuery.ts

import { useRouter } from 'next/router';

/** Returns the page query, or null if the page is not yet hydrated. */
export default function useQuery() {
  const router = useRouter();

  const hasQueryParams =
    /\[.+\]/.test(router.route) || /\?./.test(router.asPath);
  const ready = !hasQueryParams || Object.keys(router.query).length > 0;

  if (!ready) return null;

  return router.query;
}
```

* It prevents you from shooting yourself in the foot by accessing `query.foo` before the query is ready (you would get an exception since `query` is `null`)

* Works as expected when the page does not have any query params (https://github.com/vercel/next.js/issues/8259#issuecomment-544912889 does not)

* Works as expected when the URL ends with "?" (https://github.com/vercel/next.js/issues/8259#issuecomment-613472952 does not)

**Comment by zurfyx** [comment]
_My 2¢ as a non-Next.js guru. Please, correct me if I'm mistaken._ AFAIK none of the solutions seem ideal because it defeats the purpose of SSR. While the proposed solutions seem to provide a fix to the hydrated version being the same as the server-side rendered, the server-side version can't be said to be accurate according to what the user/Google-bot would eventually see.

Let's take the following as an example: default `npx create-next-app` with the following `index.js`

```
import { useEffect, useState } from "react";
import { useRouter } from "next/router";

// Don't think this is correct - refer to https://github.com/vercel/next.js/issues/8259#issuecomment-544912889
function isRouterReady(router) {
  return router.asPath !== router.route;
}

export default function Home() {
  const router = useRouter();
  const [number, setNumber] = useState(0);

  useEffect(() => {
    if (isRouterReady(router)) {
      setNumber(router.query.number ?? 0);
    }
  }, [router]);

  return <>?number={number}</>;
}
```

On `http://localhost:3000/?number=123`, server-side rendered (tested disabling JS) it will return `?number=0`, client-side it will first render `?number=0`, later `?number=123`

This means that either
1. The content is likely to swap rapidly within the first render. Bad UX, more costly (in terms of modules)
2. If you choose not to render content until the Router is reader, that's still React you could have rendered on the server (faster paint time)

**Comment by jsartisan** [comment]
NextJS seems to be broken. Moving to Gatsby. Thanks NextJS for wasting my time.

**Comment by bduffany** [comment]
Hmm, I wonder if the right solution to this issue is a matter of just deprecating the `router.query` API in favor of more newcomer-friendly APIs that make it perfectly clear what their behavior will be? I was very confused by the name `router.query` simply because I expected it to "just work" since everything else in Next.js seems to "just work." I'd imagine others in this thread feel the same. 

Once you understand how Next.js really works and why the current API works the way that it does, the API is really not problematic, but for newcomers this adds unnecessary friction and learning curve, which can be really harmful to a framework's longevity. In my opinion, the person who wrote `router.query` probably had a bit of tunnel vision and did not anticipate that many users of this API would be new to concepts like SSR.

**Comment by julian-sf** [comment]
> A place that always has access to a populated query object is `useEffect`, so try accessing to it from there.

To clarify for anyone else reading this, the statement that the query object is populated in the `useEffect` hook is untrue.   Here is a minimal example that will show what I mean:

```tsx
const Index = () => {
  const { query } = useRouter();

  console.log(">>> render", query);

  useEffect(() => {
    console.log(">>> effect", query);
  }, [query]);

  return null;
};

export default Index;
```

Here is the console output:

![image](https://user-images.githubusercontent.com/50758144/88948975-9cf1b100-d260-11ea-9750-917f64cc36e7.png)

**Comment by julian-sf** [comment]
Also note, my PR (#9370) to add `isReady` to the `Router` object will be ready by this weekend.  There were some slight issues with it, but I should have that updated soon.

**Comment by julian-sf** [comment]
> _My 2¢ as a non-Next.js guru. Please, correct me if I'm mistaken._ AFAIK none of the solutions seem ideal because it defeats the purpose of SSR. While the proposed solutions seem to provide a fix to the hydrated version being the same as the server-side rendered, the server-side version can't be said to be accurate according to what the user/Google-bot would eventually see.
> 
> Let's take the following as an example: default `npx create-next-app` with the following `index.js`
> 
> ```
> import { useEffect, useState } from "react";
> import { useRouter } from "next/router";
> 
> // Don't think this is correct - refer to https://github.com/vercel/next.js/issues/8259#issuecomment-544912889
> function isRouterReady(router) {
>   return router.asPath !== router.route;
> }
> 
> export default function Home() {
>   const router = useRouter();
>   const [number, setNumber] = useState(0);
> 
>   useEffect(() => {
>     if (isRouterReady(router)) {
>       setNumber(router.query.number ?? 0);
>     }
>   }, [router]);
> 
>   return <>?number={number}</>;
> }
> ```
> 
> On `http://localhost:3000/?number=123`, server-side rendered (tested disabling JS) it will return `?number=0`, client-side it will first render `?number=0`, later `?number=123`
> 
> This means that either
> 
> 1. The content is likely to swap rapidly within the first render. Bad UX, more costly (in terms of modules)
> 2. If you choose not to render content until the Router is reader, that's still React you could have rendered on the server (faster paint time)

You're looking at this wrong.  In order to get access to the query object on an SSR'ed page you would do something like the following instead:

```tsx
type SSRProps = { query: ParsedUrlQuery };

const Home = ({ query }: SSRProps) => {
  const [number, setNumber] = useState(query.number ?? 0);
  
  // ...

  return <>?number={number}</>;
};

export const getServerSideProps: GetServerSideProps = async ({ query }): Promise<{ props: SSRProps }> => {
  return { props: { query } };
};

export default Home;
```

Using that example the page would always be SSR'ed and the query object will be populated on THE FIRST render, so you will always have your number object (if it's set) on the first pass through this function.

**Comment by yanivfranco** [comment]
This is a major functionality issue with nextRouter + static page optimization.
This should at least be well documented, to warn other developers (like me) that spent few of my best hours on that issue.
Anyway,  @julian-sf , hopefully we will see your change in the next version! Shout out if any help is needed.

**Comment by jonathanredford** [comment]
Is there any drawback in doing this?
```
export async function getServerSideProps({query}) {
    return {
        props: {query}, // will be passed to the page component as props
    }
}
```

**Comment by mysticaltech** [comment]
I'm like a Next newbies Newbie, now I'm just brutally parsing `route.asPath` to get the query params. Lost many days because of this issue, this needs fixing.

**Comment by julian-sf** [comment]
I am so sorry I've been so slow. Things have been crazy at work (launching major new site to production). And I'm in the middle of moving during a pandemic.

The initial approach I took with the PR is incorrect, but thanks to some of the maintainers I'm aware of what the right fix would look like. I just haven't had time to get the work done myself. If anyone wants to take the context from the PR and add to it, I'm super open to that!

**Comment by idrisakmal** [comment]
I've tried running  `router.asPath !== router.route;` to do a checking when the router is ready. But the result I get on both initial and second rerender is the same. Its already false even on first load. now I've been tinkering this for hours and finally just use `window.location.search` to get the query params on first load. I dont want to use `getServerSideProps` because i still want my page to be statically rendered.

**Comment by bduffany** [comment]
@idrisakmal See https://github.com/vercel/next.js/issues/8259#issuecomment-665684705 for a fix that works.

**Comment by idrisakmal** [comment]
@bduffany Hey thanks for the solution. I hoist it up to the `_app.js` level because an empty query params is also a condition for my use ie;

```
if (isEmpty(query)) router.push(`?state=signin`, undefined, { shallow: true });
else 
 // use the query for something else
```

Thanks for your help but until there's a solution to get a ready router I'm sticking to `window.location.search` for now

**Comment by ZiiMakc** [comment]
But what should i do if i need query on SSR?

**Comment by ZiiMakc** [comment]
@bduffany i get error:
`Did not expect server HTML to contain a <div> in <header>.`

if i use conditions in render:
`{query && (<div>123</div>)}`

**Comment by mycolaos** [comment]
Facing the same problem I tried to use `getServerSideProps` to read the query in SSR. It automatically makes the `useRouter` have the correct query from the very first render, i.e. there's no empty `{}` query at the start.

1. Is it consistent behavior? I didn't find any docs on that.
2. Can it be used as the solution to the problem?
3. Why SSR just doesn't always pass the router state to the client?

**Comment by morhogg** [comment]
I have faced the same problem and published a package with a workaround

It's a tiny package that wrap the next.js router to provide the router.query and router.pathname on client-side first render

https://www.npmjs.com/package/use-client-router

**Comment by timneutkens** [comment]
> I have faced the same problem and published a package with a workaround
> 
> It's a tiny package that wrap the next.js router to provide the router.query and router.pathname on client-side first render
> 
> [npmjs.com/package/use-client-router](https://www.npmjs.com/package/use-client-router)

This will very likely cause hydration mismatches causing React to throw away the pre-rendered HTML causing performance issues and jumps/issues with the pre-rendered HTML.

**Comment by AlexandraKlein** [comment]
My query params are still not showing up in production when `router.isReady` is `true`. I'm only receiving a value for `router.query.slug`.

**Comment by jackguoAtJogg** [comment]
This could be a good solution: https://github.com/morhogg/use-client-router

**Comment by timneutkens** [comment]
@jackguoAtJogg it'll break hydration causing pages to get into a broken state so certainly would not recommend it.

**Comment by rohanx96** [comment]
As a workaround, I created a pageQuery hook that parses the utm params manually and appends them to `router.query`. This makes sure that the utm params are available on the first render on client where the hook is used.

```
  const router = useRouter();
  const query = useRef({});
  // Parse query from URL - this avoids delay in query params by next js router
  useEffect(() => {
    const urlQuery = new URLSearchParams(router.asPath.split('?')[1]);
    Array.from(urlQuery.entries()).forEach(([key, value]) => {
      query.current[key] = value;
    });
  }, [router.asPath]);
  query.current = { ...query.current, ...router.query };
  return query.current;
```

**Comment by balazsorban44** [comment]
This issue has been automatically locked due to no recent activity. If you are running into a similar issue, please create a new issue with the steps to reproduce. Thank you.
